torch/autograd/profiler.py

```diff
@@ -339,37 +391,36 @@ def __missing__(self, key):
 ################################################################################
 # CPU checkpoints
 
-Record = namedtuple('Record', ['name', 'timestamp', 'kind'])
-
-
 def parse_cpu_trace(thread_records):
     next_id = 0
-    start_time = None
+    start_record = None
     functions = []
-    function_stack = []
+    record_stack = []
     string_table = StringTable()
-    for r in itertools.chain(*thread_records):
-        record = Record(*r)
-        if record.name == '__start_profile':
-            start_time = record.timestamp
-        if record.kind == 'mark':
+    for record in itertools.chain(*thread_records):
+        if record.name() == '__start_profile':
+            start_record = record
+        if record.kind() == 'mark':
             continue
-        elif record.kind == 'push':
-            function_stack.append(FunctionEvent(
-                id=next_id, name=string_table[record.name], start=record.timestamp, end=record.timestamp))
+        elif record.kind() == 'push':
+            record_stack.append((next_id, record))
             next_id += 1
-        elif record.kind == 'pop':
-            function_stack[-1].end = record.timestamp
-            functions.append(function_stack.pop())
-
-    # Normalize times
-    if start_time is None:
-        raise RuntimeError('Malformed profile: no start marker')
-    for event in functions:
-        event.start -= start_time
-        event.end -= start_time
-
-    functions.sort(key=lambda evt: evt.start)
+        elif record.kind() == 'pop':
+            function_id, start = record_stack.pop()
+            cpu_interval = Interval(
+                start_record.cpu_elapsed_us(start),
+                start_record.cpu_elapsed_us(record))
+            cuda_interval = None
+            if start_record.has_cuda() and record.has_cuda():
```

Comment:
I am not sure, actually.The threads access use_cuda and use_nvprof without locks, so I don't have a great mental model of what is intended here. I can add an assert in any case though.

Reply:
The mental model I had was that you have a single thread that enables and disables the profiler, and you don't change the flags in the middle. This would mean that either all or no events have cuda times.

Reply:
What happens if a thread launched before or during the profiling enable outlives the end of profiling? It is going to keep push events and will race with the use_cuda flags.

Reply:
Why is it going to keep pushing events? Flag that says whether the profiling is enabled is global so it should stop queueing events once it's off, even if it outlives the profiled scope (e.g. engine threads already do this).